sentry/src/main/java/io/sentry/SentryTracer.java

```diff
@@ -0,0 +1,241 @@
+
+@ApiStatus.Internal
+public final class SentryTracer implements ITransaction {
+  private final @NotNull SentryId eventId = new SentryId();
+  private final @NotNull Span root;
+  private final @NotNull List<Span> children = new CopyOnWriteArrayList<>();
+  private final @NotNull IHub hub;
+  private final @NotNull Contexts contexts = new Contexts();
+  private @Nullable Request request;
+  private @NotNull String name;
+
+  public SentryTracer(final @NotNull TransactionContext context, final @NotNull IHub hub) {
+    Objects.requireNonNull(context, "context is required");
+    Objects.requireNonNull(hub, "hub is required");
+    this.root = new Span(context, this, hub);
+    this.name = context.getName();
+    this.hub = hub;
+  }
+
+  public @NotNull List<Span> getChildren() {
+    return children;
+  }
+
+  public @NotNull Date getStartTimestamp() {
+    return this.root.getStartTimestamp();
+  }
+
+  public @Nullable Date getTimestamp() {
+    return this.root.getTimestamp();
+  }
+
+  /**
+   * Starts a child Span with given trace id and parent span id.
+   *
+   * @param parentSpanId - parent span id
+   * @param operation - span operation name
+   * @param description - span description
+   * @return a new transaction span
+   */
+  @NotNull
+  ISpan startChild(
+      final @NotNull SpanId parentSpanId,
+      final @NotNull String operation,
+      final @Nullable String description) {
+    final ISpan span = startChild(parentSpanId, operation);
+    span.setDescription(description);
+    return span;
+  }
+
+  /**
+   * Starts a child Span with given trace id and parent span id.
+   *
+   * @param parentSpanId - parent span id
+   * @return a new transaction span
+   */
+  @NotNull
+  private ISpan startChild(final @NotNull SpanId parentSpanId, final @NotNull String operation) {
+    Objects.requireNonNull(parentSpanId, "parentSpanId is required");
+    Objects.requireNonNull(operation, "operation is required");
+    final Span span = new Span(root.getTraceId(), parentSpanId, this, operation, this.hub);
+    this.children.add(span);
+    return span;
+  }
+
+  @Override
+  public @NotNull ISpan startChild(final @NotNull String operation) {
+    return root.startChild(operation);
+  }
+
+  @Override
+  public @NotNull ISpan startChild(
+      final @NotNull String operation, final @Nullable String description) {
+    return root.startChild(operation, description);
+  }
+
+  @Override
+  public @NotNull SentryTraceHeader toSentryTrace() {
+    return root.toSentryTrace();
+  }
+
+  @Override
+  public void finish() {
+    this.finish(this.getStatus());
+  }
+
+  @Override
+  public void finish(@Nullable SpanStatus status) {
+    if (!root.isFinished()) {
+      root.finish(status);
+      hub.configureScope(
+          scope -> {
+            scope.withTransaction(
+                transaction -> {
+                  if (transaction == this) {
+                    scope.clearTransaction();
+                  }
```

Comment:
I guess we are introducing the same bug here, how do we know we are clearing the right transaction? and if its another one? we used to do an equal check, we need to find a way here.

Reply:
Since we are inside `withTransaction`, does the lock prevent from setting any transaction during this operation?